Better handling of distraction free mode state

MoritzBrueckner opened the pull request:
Previously, the distraction mode was active when the sidebar wasn't visible, so the addon didn't work well for users that have their sidebar hidden.
Now, the current state is saved inside the window's settings, so the state is independent of the user's configuration and yet gets saved between sessions. The long name of the setting was chosen on purpose to prevent possible name collisions.

MoritzBrueckner committed:
Better handling of distraction mode state

Previously, the distraction mode was active when the sidebar wasn't visible, so the addon didn't work well for users that have their sidebar hidden.
Now, the current state is saved inside the window's settings, so the state is independent of the user's configuration and yet gets saved between sessions. The long name of the setting was chosen on purpose to prevent possible name collisions.

jrappen commented:
Actually Sublime uses the Session file to determine which mode it is in, either "normal" or "Distraction Free".

`./Sublime Text 3/Local/Session.sublime-session`

I'd rather store the DistractionFreeWindow mode state there, too.

MoritzBrueckner commented:
Thank you, I didn't know that (actually this is the first time I'm working with the sublime API). Do you think it is safe to open it? I'm not sure if this will conflict with sublime text itself and I didn't find a way to store values in this file via API.

If your method is safe, it is definitely better, as this approach stores the value for the session and not for a window (otherwise project changes could lead to mode changes).

jrappen commented:
@MoritzBrueckner That file is a plain text JSON file. While Sublime is running, it uses a copy of that found in the same folder named `Auto Save Session.sublime-session`.

MoritzBrueckner commented:
And changes to that file don't get overwritten by Sublime? If no, I will try to implement this, but not before Sunday unfortunately, I'm very busy this week.

Thank you for your suggestions and help :)

jrappen commented:
@MoritzBrueckner also see discussion and relevant links in #26

MoritzBrueckner committed:
Distraction free mode state now in session file

Now the distraction free mode gets saved between sessions, between projects and even after switching windows. Should fix issue #26.

MoritzBrueckner commented:
After a lot of thinking, testing, and doubts, I finally found a solution! Now the mode gets saved between sessions, between projects and even after switching windows.

As you deleted some comments in the code in my last PR, I think it's better to explain some things here :)
- **_Why `on_activated()`?_**
I tested other callbacks, but they didn't work in every case. `on_activated()` triggers too often, but I think it's the only safe way (currently).
- **_Why the `toggle` argument instead of a dedicated WindowCommand?_**
I needed a way to apply the changes of the dfm WindowCommand without changing the state. This is also possible with another WindowCommand which applies the changes and this command would then get called after switching the state. But this would lead to another command in the command palette, so it's a bad idea.

This PR should fix #26

Are the new changes ok for you? I had to merge in the master branch because of a merge conflict.

MoritzBrueckner committed:
Merge branch 'master' into dfm_state

jrappen commented on .sublime/commands/Default.sublime-commands:
```diff
@@ -2,6 +2,7 @@
 	{
 		"caption": "DistractionFreeWindow: Toggle Mode",
 		"command": "distraction_free_window"
+        "args": {"toggle": true}
```
missing comma in previous line, wrong indent

MoritzBrueckner replied:
Sublime uses tabs instead of spaces for its own file formats, I didn't realize that. Looked right on my machine. Fixed now, thank you.

jrappen commented on main.py:
```diff
@@ -1,36 +1,57 @@
     def _status_msg(self, msg):
         self.window.status_message('Distraction Free Window: {}'.format(msg))
-
+        
```
redundant whitespace

MoritzBrueckner replied:
Strange, I don't have that whitespace and it doesn't show up in git diff either... Sublime actually deletes all trailing whitespace during file saving.

MoritzBrueckner replied:
Ah, I've found it, didn't pull before push and the whitespace got introduced by the merging of the master branch into my branch. Also fixed now.

jrappen commented on main.py:
```diff
@@ -1,36 +1,57 @@
-    def run(self):
+    def run(self, toggle=True):
+        packages_path = sublime.packages_path()
+        session_file = os.path.join(os.path.dirname(packages_path), 'Local', 'Session.sublime_session')
```
Not sure whether we should take the `Session` or rather the `Auto Save Session` file here. Is there a reason for your pick? I'll have to test this first on a few different setups.

MoritzBrueckner replied:
The reason was that I thought it might be better in order to avoid read/write conflicts. I don't know which one is better, but this one works for me.

Looking forward for your test results :)

jrappen commented on main.py:
```diff
@@ -83,11 +104,20 @@ def run(self):
+                pass
+
+            try:
+                with open(session_file, "w") as file:
```
```python
with os.open(session_file, mode='w', newline='\n') as file:
```

MoritzBrueckner replied:
I think you mean open() instead of os.open(), else this error is raised:
TypeError: Required argument 'flags' (pos 2) not found

Added the newline param in open(), thank you.

jrappen commented on main.py:
```diff
@@ -83,11 +104,20 @@ def run(self):
+                with open(session_file, "w") as file:
+                    file.write(sublime.encode_value(session_settings, pretty=True))
+            except FileNotFoundError:
+                self._status_msg('Error: Session file not found! Mode was not saved!')
```
```python
if session_settings is not None:
    with os.open(session_file, mode='x', newline='\n') as file:
        file.write(sublime.encode_value(session_settings, pretty=True))
```

would create the file, although it doesn't make sense to do that here. Just a reference for interested reviewers.

MoritzBrueckner replied:
Why not?

MoritzBrueckner committed:
Small adjustments for PR #32

MoritzBrueckner committed:
Merge branch 'dfm_state' of https://github.com/MoritzBrueckner/DistractionFreeWindow into dfm_state

MoritzBrueckner committed:
Remove trailing whitespace

MoritzBrueckner commented:
@jrappen As it seems to me, after the latest Sublime Text update the method with writing to session files doesn't work anymore. Or it also didn't work before and I wasn't realizing that and I'm just too stupid to find the mistake.

Why not save the state in the package settings? Using session files seems overly complicated to me. If you think package settings are a good idea, I can create a new, clean PR for this and we close this one.

jrappen commented:
@MoritzBrueckner The following is happening currently when using your code:

* `Session.sublime-session` file is copied to `Auto Save Session.sublime-session` upon ST startup.
* You code reads from and writes to the `Session.sublime-session` file.
* When ST is closed, ST overwrites the `Session.sublime-session` file with your changes from the `Auto Save Session.sublime-session` file and deletes the `Auto Save Session.sublime-session` file when done.
  * When ST crashes, it ignores the `Session.sublime-session` upon the next startup if the `Auto Save Session.sublime-session` file is found.

Changing:

https://github.com/aziz/DistractionFreeWindow/blob/a0247df037871df745651cb3d7f12eb5ea9136e1/main.py#L24-L35

to:

```python
session_file = os.path.join(os.path.dirname(packages_path), 'Local', 'Auto Save Session.sublime_session')
```

should work as expected. Could you test the suggested change on your end?

jrappen commented:
@MoritzBrueckner On another note we should track DistractionFreeWindow mode per window as toggling it is per window as well.

In the JSON session files there's a `"windows": [ ... ]` array where every window can be accessed via the `"windows_id"` value.

Compare [`sublime.window.id()`](https://www.sublimetext.com/docs/3/api_reference.html#sublime.Window).

MoritzBrueckner commented:
> Could you test the suggested change on your end?

Using "Auto Save Session.sublime-session" seems to work.

> In the JSON session files there's a "windows": [ ... ] array where every window can be accessed via the "windows_id" value.

In the session file, the (only) window id is 404, whereas w.id() is 2. I don't get this to work.

jrappen commented:
@MoritzBrueckner ok, thanks for the update. Will test on my side as soon as I get back from visiting friends abroad.

MoritzBrueckner commented:
@jrappen Any news on this?
